src/Symfony/Component/PropertyInfo/Extractor/PhpStanExtractor.php

```diff
@@ -63,103 +68,111 @@ public function __construct(?array $mutatorPrefixes = null, ?array $accessorPref
                 continue;
             }
 
-            foreach ($this->phpStanTypeHelper->getTypes($tagDocNode->value, $nameScope) as $type) {
-                switch ($type->getClassName()) {
-                    case 'self':
-                    case 'static':
-                        $resolvedClass = $class;
-                        break;
-
-                    case 'parent':
-                        if (false !== $resolvedClass = $parentClass ??= get_parent_class($class)) {
-                            break;
-                        }
-                        // no break
-
-                    default:
-                        $types[] = $type;
-                        continue 2;
+            try {
+                $types[] = $this->stringTypeResolver->resolve((string) $tagDocNode->value->type, $typeContext, $backwardCompatible);
+            } catch (UnsupportedException) {
+                if ($backwardCompatible && 'void' === (string) $tagDocNode->value->type) {
+                    return Type::void();
+                }
+            } catch (LogicException $e) {
+                // handle "parent" type without existing parent
+                if ($backwardCompatible && 'parent' === (string) $tagDocNode->value->type) {
+                    return Type::object('parent');
                 }
 
-                $types[] = new Type(Type::BUILTIN_TYPE_OBJECT, $type->isNullable(), $resolvedClass, $type->isCollection(), $type->getCollectionKeyTypes(), $type->getCollectionValueTypes());
+                throw $e;
             }
         }
 
-        if (!isset($types[0])) {
+        if (null === $type = $types[0] ?? null) {
             return null;
         }
 
-        if (!\in_array($prefix, $this->arrayMutatorPrefixes, true)) {
-            return $types;
+        if (!\in_array($prefix, $this->arrayMutatorPrefixes)) {
+            return $type;
         }
 
-        return [new Type(Type::BUILTIN_TYPE_ARRAY, false, null, true, new Type(Type::BUILTIN_TYPE_INT), $types[0])];
+        return Type::list($type);
     }
 
-    public function getTypesFromConstructor(string $class, string $property): ?array
+    /**
+     * @deprecated since Symfony 7.1, use "getType" instead
+     */
+    public function getTypes(string $class, string $property, array $context = []): ?array
+    {
+        trigger_deprecation('symfony/property-info', '7.1', 'The "%s()" method is deprecated, use "%s::getType()" instead.', __METHOD__, self::class);
+
+        return BackwardCompatibilityHelper::convertTypeToLegacyTypes($this->getType($class, $property, $context, true));
+    }
+
+    public function getTypeFromConstructor(string $class, string $property): ?Type
     {
+        // BC layer to allow unconsistent legacy type extraction
+        // Can be removed as soon as PhpStanExtractor::getTypesFromConstructor() is removed (8.0).
+        $backwardCompatible = \func_get_args()[2] ?? false;
```

Comment:
I feel like all these $backwardCompatible variables might make cross-majors compatibility (7.4x8.0) difficult. Can't we keep and run the legacy code in getTypesFromConstructor instead, and use the same approach everywhere?